include/jemalloc/internal/psset.h

```diff
@@ -61,8 +61,15 @@ struct psset_s {
 	hpdata_age_heap_t pageslabs[PSSET_NPSIZES];
 	bitmap_t bitmap[BITMAP_GROUPS(PSSET_NPSIZES)];
 	psset_stats_t stats;
-	/* Slabs with no active allocations. */
-	hpdata_empty_list_t empty_slabs;
+	/*
+	 * Slabs with no active allocations, but which are allowed to serve new
+	 * allocations.
+	 */
+	hpdata_empty_list_t empty;
+	/* Slabs which are available to be purged. */
+	hpdata_purge_list_t to_purge;
+	/* Slabs which are available to be hugified. */
+	hpdata_hugify_list_t to_hugify;
```

Comment:
It's currently possible for an hpdata to be a candidate both for purging (which depends on shard-global properties) and for hugifying (which depends on hpdata-local properties), so they need to be separate lists.[1]

I think empty and to_hugify could share space; but we're not particularly strapped for space in the hpdata, so I figured we'd just do the convenient thing.

[1] This is a little unintuitive, but I think it makes sense; suppose you have two otherwise indistinguishable hugepage-regions,[2] both of which satisfy the hugification criterion, but which you can't hugify either because of dirty page limits. You want to purge one to allow you to hugify the other.

[2] Where a hugepage region is defined as a hugepage-aligned, hugepage-sized region of address space, regardless of whether or not it's actually backed by a hugepage.[3]

[3] That was indeed a footnote within a footnote.

Reply:
I see. Also agree that we need to define some of the terms, so that we can talk sanely w/o nested footnotes...